Quickly prioritize candidate disease-causing WES variants based on functional annotations



Whole exome sequencing (WES) has been commonly used in clinical settings to help with diagnosis and explore the genetic basis of rare mendelian diseases. Given its advantages over gene panel testing, WES still has several hurdles that limit physicians or clinical geneticists' application of such techonology.
Some of the hurdles identified:
  - Prioritization: 10,000 + variants identified per-patient, most of them 	have unknown significance. 
  - Interpretability

This pipeline is specifically designed for physicians and clinical interpretation of WES variants. 

Currently, the python code is for specific input files. You will need to tweak the codes to read your own files. We will keep updating the code to make it easily accessible to everyone. 

The current python script takes individual's annotated genomic variant file from WGSA (https://sites.google.com/site/jpopgen/wgsa) and dbNSFP (https://sites.google.com/site/jpopgen/dbNSFP) as input. Please refer to these site in terms of how to generate such input file. 


